rest/server.go

```diff
@@ -0,0 +1,98 @@
+package rest
+
+import (
+	"awesomeProject3/service"
+	"context"
+	"encoding/json"
+	"github.com/go-chi/chi/v5"
+	"log"
+	"net/http"
+	"time"
+)
+
+type container struct {
+	Getter string `json:"getter"`
+	Value  string `json:"value"`
+}
+
+type SecretRestAPI struct {
+	ssService        service.SecretService
+	shutdown         chan struct{}
+	shutdownComplete chan struct{}
+	port             string
+}
+
+func NewSecretRestAPI(sr service.SecretService, portnumber string) *SecretRestAPI {
+	return &SecretRestAPI{port: portnumber, ssService: sr, shutdown: make(chan struct{}, 1), shutdownComplete: make(chan struct{}, 1)}
+}
+
+func (sr *SecretRestAPI) Stop() {
+	log.Println("started stop in the rest")
+	sr.shutdown <- struct{}{}
+	<-sr.shutdownComplete
+}
+
+func (sr *SecretRestAPI) Start(serverCtx context.Context) error {
+	h := chi.NewRouter()
+	h.MethodFunc("GET", "/", sr.handlerGet)
+	h.MethodFunc("POST", "/", sr.handlerPost)
+
+	httpServer := &http.Server{Addr: ":" + sr.port, Handler: h}
+
+	go func() {
+		<-sr.shutdown
+		log.Println("Try to shutdown gracefully")
+		shutdownCtx, shutDownStopCtx := context.WithTimeout(serverCtx, 5*time.Second)
+		defer shutDownStopCtx()
+		err := httpServer.Shutdown(shutdownCtx)
+		if err != nil {
+			log.Println("Error trying to shutdown server. Error: ", err)
+
+		}
+
+		sr.shutdownComplete <- struct{}{}
+	}()
+
+	err := httpServer.ListenAndServe()
+	if err != nil && err != http.ErrServerClosed {
+		return err
+	}
+	log.Println("HTTP Server stopped successfully")
+
+	return nil
+}
```

Comment:
This is approximately what I was waited for. Since `Start` encapsulates `httpServer` and knows how to handle it,  so logic with shutting it down should also be placed closer to the scope where `httpServer` was defined.

`ctx` is the only object that should manipulate state of the server and its lifetime; `sync.WaitGroup` organises synchronous execution in multi goroutine environment and would allow `Start()` method to be a blocking synchronous method.

```suggestion
func (sr *SecretRestAPI) Start(ctx context.Context) error {
	h := chi.NewRouter()
	h.MethodFunc("GET", "/", sr.handlerGet)
	h.MethodFunc("POST", "/", sr.handlerPost)

	httpServer := &http.Server{Addr: ":" + sr.port, Handler: h}
	
	var wg sync.WaitGroup
	wg.Add(2)
	
	var err error
	
	go func() {
	     defer wg.Done()
	     srvErr := httpServer.ListenAndServe()
	     if srvErr != nil && srvErr != http.ErrServerClosed {
		    err = srvErr
	     }
	}()

	go func() {
	       defer wg.Done()
		<-ctx.Done()
		log.Println("Try to shutdown gracefully")
		shutdownCtx, shutDownStopCtx := context.WithTimeout(context.Background(), 5*time.Second)
		defer shutDownStopCtx()
		err := httpServer.Shutdown(shutdownCtx)
		if err != nil {
			log.Println("Error trying to shutdown server. Error: ", err)
		}
	}()

        wg.Wait()
        log.Println("HTTP Server stopped successfully")
        if err != nil {
            return err
        }
        
	return nil
}
```